src/FindEmailAddress.java

```diff
@@ -0,0 +1,362 @@
+            URI uri = new URI(rootURL);
+            this.requestDomain= uri.getHost();
+
+        }catch(java.net.URISyntaxException e)
+        {
+            e.printStackTrace();
+        }
+        this.emails = new Hashtable<String,String>();
+        this.urlHistory = new Hashtable<String, String>();
+        this.urlHistory.put(rootURL,"");
+        tasksNumber = 0;
+    }
+
+    /*
+     *  This is the function to do web crawling
+     *  it will create a thread pool
+     *  if all the tasks are done, it will stop.
+     */
+
+    public void startWebCrawl()
+    {
+        if(rootURL == null || rootURL.trim().length()<=0)
+            return;
+
+        executor = Executors.newFixedThreadPool(10);
+        executor.execute(new Handler(rootURL));
+        do{
+            try {
+                Thread.sleep(3000);
+            }catch(Exception e)
+            {
+                e.printStackTrace();
+            }
+        }while(this.tasksNumber >0);
+
+        executor.shutdown();
+    }
+
+/*
+* this inline class is for http request and email check running task
+*/
+    class Handler implements Runnable {
+        String requestURL;
+        Handler(String requestURL) {
+            this.requestURL = requestURL;
+        }
+        public void run() {
+            System.out.println("start search url :" + requestURL);
+            makeHttpReuqest(requestURL);
+        }
+    }
+
+    //need multithread
+    /*
+    * This is the function to make http call.
+    * And this function will also do the email search and same domain url search.
+    * the email address will be saved in a hashtable
+    * the urls will be saved in a vector, and a history tashtable is for filtering the urls already visited.
+    */
+    private void makeHttpReuqest(String httpURLString)
+    {
+        if(httpURLString == null || httpURLString.trim().length()<=0)
+            return;
+        BufferedReader reader = null;
+        try {
+            URL urlObj = new URL(httpURLString);
+            HttpURLConnection con = (HttpURLConnection) urlObj.openConnection();
+            con.setRequestMethod("GET");
+            con.setRequestProperty("User-Agent", "Mozilla/5.0"); //some website if did not see the user agent will take it robot and cause http call failing
+
+            reader = new BufferedReader(
+                    new InputStreamReader(con.getInputStream()));
+            String inputLine;
+
+            while ((inputLine = reader.readLine()) != null) {
+                searchEmailAndHRef(inputLine);
+            }
+
+        }catch(Exception ex)
+        {
+            System.out.println("Cannot reach url:" + httpURLString);
+        }finally{
+            if(reader != null)
+            {
+                try {
+                    reader.close();
+                }catch (Exception e)
+                {
+                    e.printStackTrace();
+                }
+                synchronized (tasksNumber) {
+                    tasksNumber --;
```

Comment:
when I ran your code against `web.mit.edu`, after it finished it sat in an infinite loop forever, because `tasksNumber` never dropped to zero. looking at this code I see one pathway where it won't get decremented:
* urlObj.openConnection throws an exception
* exception is caught and "Cannot reach url" is printed, but reader is null
* but you're only decrementing the counter here if `reader != null`

Reply:
confirmed, once I moved `tasksNumber--` out of the `if` statement, the program exits.

Reply:
Thanks, big mistake.